### PR TITLE
Fix #139: Correct allele ID found even when PRG alphabet is discontinuous

### LINKED ISSUE #139: Kmer index unit test assigns an allele ID of zero to search state
Unit test: https://github.com/iqbal-lab-org/gramtools/blob/ad47bb63a09c4b0b1f92f005e2059ce842223567/libgramtools/tests/kmer_index/test_build.cpp#L671

This function is called whilst running the above unit test: https://github.com/iqbal-lab-org/gramtools/blob/03ff4ff2309f25f5470816b3097d60ea7f0a9754/libgramtools/src/search/search.cpp#L336

Within the above function, a call to `get_allele_id(.)` returns zero.

The allele mask is generated and seems reasonable. The first step in the investigation is to generate PRG table information (SA, F, ...).

### PR DESCRIPTION
**Pull request by rffrancon**
This edge case is currently only found in tests.

This solution handles the edge case and makes gramtools more robust in the eventuality that a single PRG is analyzed in parts.

### CODE REVIEW

File: libgramtools/src/search/search.cpp [hunk 1]
```diff
@@ -308,8 +308,12 @@ SA_Interval gram::get_allele_marker_sa_interval(const Marker &site_marker_char,
     const auto start_sa_index = prg_info.fm_index.C[alphabet_rank];
 
     const auto next_boundary_marker = allele_marker_char + 1;
-    const auto max_char_in_alphabet = prg_info.fm_index.sigma - 1;
-    const bool next_boundary_marker_valid = next_boundary_marker <= max_char_in_alphabet;
+    // sigma: contains the size (=number of unique symbols) of the alphabet
+    const auto max_alphabet_char = prg_info.fm_index.comp2char[prg_info.fm_index.sigma - 1];
```
Comment (bricoletc) [inline libgramtools/src/search/search.cpp]: Nice fix. This no longer assumes that all integers between 1 and sigma are actually being used in the prg (ie prg alphabet can be **discontinuous**.) 

I think there are other places in gramtools where prgs encoded using a discontinuous alphabet would cause trouble. One example is line 310:

 `const auto next_boundary_marker = allele_marker_char + 1; `

If the prg uses for example only 5 and 23 as site markers, then from `allele_marker_char` of 6, we would assign `next_boundary_marker` to 7, which is non-existent.

I will write these similar types of problems where spotted in `TODOs` in `add_doc` branch.

**Comment by bricoletc** [comment]
Handles a case of a prg with **discontinuous alphabet**: meaning,  that not all integers from 1 to the largest integer in the prg are present in the prg.  

The case is handled for one part of the code only; to be handled more generally, more work will be needed.
